### PR TITLE
refactor: price chart organization

### PR DESCRIPTION
**Pull request by cartcrom**
## Description

Refactors PriceChart.tsx to use a `ChartModel` object that consolidates vars and utility used to render the svg. This object is passed around between components, allowing for a simpler component tree, less interdependent useMemos, and less null-checks.

Adds more unit tests for edge cases where data is stale, undefined, or an empty array.

### QA (ie manual testing)

- [ ] View a variety of different Price Charts, comparing to prod to ensure nothing has changed visually

### CODE REVIEW

File: src/components/Tokens/TokenDetails/ChartSection.tsx [hunk 1]
```diff
@@ -60,7 +60,7 @@ function Chart({
   return (
     <ChartContainer data-testid="chart-container">
       <ParentSize>
-        {({ width }) => <PriceChart prices={prices ?? null} width={width} height={436} timePeriod={timePeriod} />}
+        {({ width }) => <PriceChart prices={prices} width={width} height={392} timePeriod={timePeriod} />}
```
Comment (cartcrom) [inline src/components/Tokens/TokenDetails/ChartSection.tsx:63]: previously, PriceChart was arbitrarily adding subtracting 44px from the passed height. this oddity was leftover from previous refactor so I removed it.

**Review by JackShort — APPROVED** [review]

**Review by just-toby — APPROVED** [review]
looks clean. i left a few questions

File: src/components/Tokens/TokenDetails/Skeleton.tsx [hunk 2]
```diff
@@ -4,10 +4,10 @@ import { useParams } from 'react-router-dom'
 import styled, { useTheme } from 'styled-components'
 import { textFadeIn } from 'theme/styles'
 
+import { TokenPrice } from '../../Charts/PriceChart'
```
Comment (cbachmeier) [inline src/components/Tokens/TokenDetails/Skeleton.tsx:7]: nit: use relative pathing

**Review by cbachmeier — APPROVED** [review]
LGTM only a minor nit and a Q

File: src/components/Charts/PriceChart/index.tsx [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            getX={getX}
+            getY={getY}
+            marginTop={dimensions.marginTop}
+            curve={curve}
+            strokeWidth={2}
+            color={theme.neutral3}
+            dashed
+          />
+        ))}
+        {crosshair !== undefined ? (
```
Comment (cbachmeier) [inline src/components/Charts/PriceChart/index.tsx:174]: this used to be `crosshair !== null` checking this is intentional

File: src/components/Charts/PriceChart/ChartModel.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  const startingPrice = prices[0]
+  const endingPrice = prices[prices.length - 1]
+  const { min, max } = getPriceBounds(prices)
+
+  // x-axis scale
+  const timeScale = scaleLinear().domain([startingPrice.timestamp, endingPrice.timestamp]).range([0, dimensions.width])
+
+  // y-axis scale
+  const priceScale = scaleLinear().domain([min, max]).range([innerHeight, 0])
```
[the remark this answers is not in the document]
Comment (cartcrom) [reply src/components/Charts/PriceChart/ChartModel.ts:57]: So now we memoize the entire object this function returns, so these will only get recomputed if the prices or chart dimensions change, as desired

File: src/components/Charts/PriceChart/index.tsx (hunk 3 above)
Comment (cartcrom) [reply src/components/Charts/PriceChart/index.tsx:174]: yes, when I first wrote this code I chose null for some reason, but now I'm an undefined-maxi

File: src/components/Charts/PriceChart/index.tsx [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+import { ReactNode, useCallback, useEffect, useMemo, useState } from 'react'
+import { Info } from 'react-feather'
+import styled, { useTheme } from 'styled-components'
+import { ThemedText } from 'theme'
+import { textFadeIn } from 'theme/styles'
+import { formatUSDPrice } from 'utils/formatNumbers'
+
+import { calculateDelta, DeltaArrow, formatDelta } from '../../Tokens/TokenDetails/Delta'
+
+const CHART_MARGIN = { top: 100, bottom: 48, crosshair: 72 }
```
[the remark this answers is not in the document]
Comment (cartcrom) [reply src/components/Charts/PriceChart/index.tsx:25]: not yet, these are just hardcoded defaults for now

File: src/components/Charts/PriceChart/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import { TimePeriod } from 'graphql/data/util'
+import { render, screen } from 'test-utils/render'
+
+import { PriceChart } from '.'
+
+jest.mock('components/Charts/AnimatedInLineChart', () => ({
+  __esModule: true,
```
[the remark this answers is not in the document]
Comment (cartcrom) [reply src/components/Charts/PriceChart/index.test.tsx:7]: I think it might have to do with them being memoized? not exactly sure since I didn't write this part originally, would have to look deeper

**Review by just-toby — APPROVED** [review]
